applications/nrf5340_audio/src/modules/power_meas.c

```diff
@@ -0,0 +1,242 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/drivers/sensor.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/shell/shell.h>
+#include <zephyr/sys/util.h>
+
+LOG_MODULE_REGISTER(power_meas, CONFIG_MODULE_POWER_LOG_LEVEL);
+
+#define POWER_RAIL_NUM 4U
+
+struct power_rail_config {
+	const char *name;
+	const struct device *sensor;
+};
+
+static const struct power_rail_config rail_config[POWER_RAIL_NUM] = {
+	{"VBAT", DEVICE_DT_GET(DT_NODELABEL(vbat_sensor))},
+	{"VDD1_CODEC", DEVICE_DT_GET(DT_NODELABEL(vdd1_codec_sensor))},
+	{"VDD2_CODEC", DEVICE_DT_GET(DT_NODELABEL(vdd2_codec_sensor))},
+	{"VDD2_NRF", DEVICE_DT_GET(DT_NODELABEL(vdd2_nrf_sensor))},
+};
+static bool rail_enabled[POWER_RAIL_NUM];
+
+K_THREAD_STACK_DEFINE(meas_stack, CONFIG_POWER_MEAS_STACK_SIZE);
+static struct k_thread meas_thread;
+K_SEM_DEFINE(meas_sem, 0, 1);
+static bool meas_enabled;
+
+/**
+ * @brief Read a power rail data (voltage/current/power) and log results.
+ *
+ * @param[in] config Power rail configuration.
+ *
+ * @retval 0 on success
+ * @retval -errno on sensor fetch/get error.
+ */
+static int read_and_log(const struct power_rail_config *config)
+{
+	int ret;
+	struct sensor_value voltage, current, power;
+
+	ret = sensor_sample_fetch(config->sensor);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = sensor_channel_get(config->sensor, SENSOR_CHAN_VOLTAGE, &voltage);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = sensor_channel_get(config->sensor, SENSOR_CHAN_CURRENT, &current);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = sensor_channel_get(config->sensor, SENSOR_CHAN_POWER, &power);
+	if (ret < 0) {
+		return ret;
+	}
+
+	LOG_INF("%-10s:\t%.3fV, %06.3fmA, %06.3fmW", config->name,
+		sensor_value_to_double(&voltage),
+		sensor_value_to_double(&current) * 1000.0f,
+		sensor_value_to_double(&power) * 1000.0f);
+
+	return 0;
+}
+
+/** @brief Measurement thread */
+static void meas_thread_fn(void *dummy1, void *dummy2, void *dummy3)
+{
+	ARG_UNUSED(dummy1);
+	ARG_UNUSED(dummy2);
+	ARG_UNUSED(dummy3);
+
+	while (1) {
+		if (!meas_enabled) {
+			k_sem_take(&meas_sem, K_FOREVER);
+		}
+
+		for (size_t i = 0U; i < POWER_RAIL_NUM; i++) {
+			int ret;
+
+			if (!rail_enabled[i]) {
+				continue;
+			}
+
+			ret = read_and_log(&rail_config[i]);
+			if (ret < 0) {
+				LOG_ERR("Could not read %s", rail_config->name);
+			}
+		}
+
+		k_msleep(CONFIG_POWER_MEAS_INTERVAL_MS);
```

Comment:
I would like to share a thought, which is not necessarily a blocker:
 
The way this module has been rewritten, a timer keeps track of when to read the INAs.
This is not ideal, as the timers in the INAs themselves should keep track of this. An average measurement is started, then INA x will assert the alert pin. We should then trigger an interrupt here, read which INA has finished the measurement and then read out the values.

Reply:
If, on the other hand, this diver is only configured for single-triggering measurements (as is done in `sensor_sample_fetch`):

This seems to void the INA function of averaging over N measurements. This reduces the accuracy and gives more I2C traffic.
Also, how do we know that the driver know that the INA has finished when we call `sensor_channel_get`? I have not dug into the lower level drivers yet, but it may be that if the measurement duration is long, we can risk reading out values before the measurement has finsihed.

Reply:
note: alert was not working at all, at least on my board (I could observe warnings). I actually get similar readings with current setup.

Reply:
Note that I have copied all the existing configurations, do you suggest to increase polling interval or reduce averaging time?

Reply:
The configs are fine. What I suggest is that this line: `k_msleep(CONFIG_POWER_MEAS_INTERVAL_MS);` is replaced by the actual configuration set in DTS

Reply:
To elaborate: `config POWER_MEAS_INTERVAL_MS` only affects the prints and not the actual power meas interval. This KConfig can be removed and rather use pseudo:
`k_msleep((ina230.config.conv_time_curr * + ina230.config.conv_time_voltage)*ina230.config.avg_mode),`

Reply:
Updated to 8500ms, and added details on how the value should be configured.